react-custom-hooks/client-ts/src/components/useTodos.ts

```diff
@@ -0,0 +1,135 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- Remove me */
+import { useEffect, useState } from 'react';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export type UseTodosValues = {
+  isLoading: boolean;
+  error?: unknown;
+  todos?: Todo[];
+  addTodo: (todo: UnsavedTodo) => Promise<void>;
+  toggleCompleted: (todoId: number) => Promise<void>;
+};
+/**
+ * Manages the Todos by reading from and writing to the backend API,
+ * using the API functions below.
+ */
+export function useTodos(): UseTodosValues {
+  const [todos, setTodos] = useState<Todo[]>();
+  const [error, setError] = useState<unknown>();
+  useEffect(() => {
+    /* TODO: If `todos` hasn't been defined yet, read the items from the API
+     * and set them into the `todos` state.
+     */
+    async function fetchTodos() {
+      try {
+        const todosData = await readTodos();
+        setTodos(todosData);
+      } catch (error) {
+        setError(error);
+      }
+    }
+    if (!todos) {
+      fetchTodos();
+    }
+  }, [todos]);
+
+  async function addTodo(newTodo: UnsavedTodo): Promise<void> {
+    /* TODO: Call the API function that creates a Todo item.
+     * When the promise returned by that function resolves, update the `todos` state.
+     * Note that it is critical that you pass a _new_ array. Do not modify the `todos` array.
+     */
+    try {
+      const createdTodo = await createTodo(newTodo);
+      const updatedTodos = [...(todos ?? []), createdTodo];
```

Comment:
The `??` is safe, but this function won't ever be called if `todos` is undefined.

TypeScript doesn't know that, though. The way I handle this is to check it for undefined at the top of the function:
```
async function addTodo(newTodo: UnsavedTodo): Promise<void> {
  if (!todos) throw new Error('todos was undefined!');
  try {
   ... etc.
```

Then I don't need to do the `??` trick or other checks for undefined. It makes the code more robust since TypeScript now knows the proper type of the variable.